File: 2024/03/ec.pl
#!/usr/bin/perl
use strict;
use warnings;
use v5.30;
use lib "../../lib-perl";
no warnings 'portable';
use Helpers qw/:all/;

sub parse {
  my ($in) = @_;
  chomp $in;
  $in =~ s/\./0/g;
  my $s = ($in =~ y/#/1/);
  my $m = [map {[split //, $_]} split /\n/, $in];
  my $h = @$m;
  my $w = @{$m->[0]};
  return [$m, $w, $h, $s];
}

my $i1 = parse(input(1));
my $a1 = dig($i1);
my $i2 = parse(input(2));
my $a2 = dig($i2);
my $i3 = parse(input(3));
my $a3 = dig($i3, 1);
print "Part 1: ", $a1, "\n";
print "Part 2: ", $a2, "\n";
print "Part 3: ", $a3, "\n";

sub dig {
  my ($in, $diagonal) = @_;
  my @offsets =
    $diagonal
    ? ([-1, -1], [0, -1], [1, -1], [-1, 0], [1, 0], [-1, 1], [0, 1], [1, 1])
    : ([0, -1], [0, 1], [-1, 0], [1, 0]);
  my $a1 = $in->[3];
  my $done;
  my $n = 2;
  while (!$done) {
    $done = 1;
    for my $y (1 .. $in->[2] - 2) {
    SQ:
      for my $x (1 .. $in->[1] - 2) {
        my $c = $in->[0]->[$y]->[$x];
        if ($c != $n - 1) {
          next;
        }
        for my $o (@offsets) {
          my $c = $in->[0]->[$y + $o->[1]]->[$x + $o->[0]];
          if ($c < $n - 1) {
            next SQ;
          }
        }
        $in->[0]->[$y]->[$x] = $n;
        $a1++;
        $done = 0;
      }
    }
    $n++;
  }
  # my $m = join "", (map {(join "", @$_) . "\n"} @{$in->[0]});
  # print $m;
  return $a1;
}

File: 2024/03/ec.go
package main

import (
	"fmt"
	"strings"

	. "github.com/beanz/everybodycodes/lib-go"
)

func parts(i1, i2, i3 []byte) (int, int, int) {
	p1 := parse(i1)
	p2 := parse(i2)
	p3 := parse(i3)
	return dig(p1, false), dig(p2, false), dig(p3, true)
}

func dig(p *parsed, diagonal bool) int {
	offsets := [][]int{{-1, -1}, {0, -1}, {1, -1}, {-1, 0}, {1, 0}, {-1, 1}, {0, 1}, {1, 1}}
	if !diagonal {
		offsets = [][]int{{0, -1}, {-1, 0}, {1, 0}, {0, 1}}
	}
	res := p.s
	var done bool
	n := byte(2)
	for !done {
		done = true
		for y := 1; y < p.h-1; y++ {
		SQ:
			for x := 1; x < p.w-2; x++ {
				if p.m[y*p.w+x] != n-1 {
					continue
				}
				for _, offset := range offsets {
					if p.m[(y+offset[1])*p.w+(x+offset[0])] < n-1 {
						continue SQ
					}
				}
				p.m[y*p.w+x] = n
				res++
				done = false
			}
		}
		n++
	}
	return res
}

func main() {
	p1, p2, p3 := parts(Input(1, ""), Input(2, ""), Input(3, ""))
	fmt.Printf("Part 1: %v\nPart 2: %v\nPart 3: %v\n", p1, p2, p3)
}

type parsed struct {
	m    []byte
	w, h int
	s    int
}

func parse(in []byte) *parsed {
	w, h := 0, 0
	j := 0
	s := 0
	for i := range len(in) {
		switch in[i] {
		case '\n':
			h++
			w = j
			j = 0
		case '.':
			in[i] = 0
		case '#':
			s++
			in[i] = 1
		}
		j++
	}
	return &parsed{in, w, h + 1, s}
}

func (p *parsed) String() string {
	var sb strings.Builder
	for y := range p.h {
		for x := range p.w - 1 {
			sb.WriteByte(p.m[y*p.w+x] + '0')
		}
		sb.WriteByte('\n')
	}
	return sb.String()
}
